### PR TITLE
Move optimization to pytrees

### PR DESCRIPTION
**Pull request by janosg**
## To-Do

- [x] Refactor constraints code to make it less DataFrame dependent
- [x] Write code to process loc, query, selector and index fields of constraints
- [x] Disentangle the transformations we do to parameters and outputs of criterion functions and create converters for each step.
    - [x] flattening / tree_transformations
    - [x] constraints / space_transformations
    - [x] scaling
- [x] Write a combined converter
- [x] Use new functionality in optimization
- [x] Use new functionality in estimation
- [x] Add fast paths
- [x] Adjust documentation

### CODE REVIEW

**Review by timmens — APPROVED** [review]
Amazing PR! I really like the new converter modules ---very clean. I only found some typos in the docs.

File: src/estimagic/parameters/parameter_groups.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+def get_params_groups(flat_params):
+    return ["a"] * len(flat_params.names)
```
[the remark this answers is not in the document]
Comment (janosg) [reply src/estimagic/parameters/parameter_groups.py:2]: Currently it's just a placeholder. It should create groups of parameters that are visualized in one plot in the dashboard. I wrote a to-do in #313 that it needs to be actually implemented. But did not want to invest much time before we have our dashboard brainstorming session.

File: src/estimagic/optimization/scipy_optimizers.py [hunk 2]
```diff
@@ -441,13 +441,7 @@ def process_scipy_result(scipy_results_obj):
 
 
 def get_scipy_bounds(lower_bounds, upper_bounds):
```
[the remark this answers is not in the document]
Comment (janosg) [reply src/estimagic/optimization/scipy_optimizers.py:443]: Yes, that's fine. But the functions should both be private. Will fix that.

[the remark this answers is not in the document]
Comment (janosg) [reply src/estimagic/optimization/scipy_optimizers.py:443]: Done.

File: docs/source/how_to_guides/optimization/index.rst [hunk 3]
```diff
@@ -8,12 +8,11 @@ Optimization
...
-   how_to_select_parameters.ipynb
    how_to_specify_constraints
    how_to_use_logging
    how_to_use_the_dashboard
    how_to_handle_errors_during_optimization
-   how_to_choose_optimizer
    how_to_scale_optimization_problems
    how_to_do_multistart_optimizations.ipynb
    how_to_benchmark_optimization_algorithms.ipynb
+   visualizing_an_optimization_problem.ipynb
```
[the remark this answers is not in the document]
Comment (janosg) [reply docs/source/how_to_guides/optimization/index.rst:18]: Done.

**Pull request merged**